File: Cuarto/Programación Distribuida/java_concurrencia/src/Ejercicio5.java
public class Ejercicio5 {
    public static class WriterThread extends Thread {
        String text;
        int id;

        public WriterThread(String text, int id){
            this.text = text;
            this.id = id;
        }

        @Override
        public void run() {
            for (int i = 0; i < 5; i++) {
                if (turno && id == 1){
                    System.out.print(text);
                    turno = false;
                    notifyAll();
                    try {
                        wait();
                    } catch (InterruptedException e) {
                        throw new RuntimeException(e);
                    }
                } else if(!turno &&  id == 2) {
                    System.out.print(text);
                    turno = true;
                    notifyAll();
                    try {
                        wait();
                    } catch (InterruptedException e) {
                        throw new RuntimeException(e);
                    }
                }
            }
        }
    }
    static boolean turno = true;

    public static void main(String[] args) {
        Thread t_hola = new WriterThread("Hola", 1);
        Thread t_mundo = new WriterThread(" Mundo\n", 2);
        t_hola.start();
        t_mundo.start();
    }
}

File: Cuarto/ProgramaciónDistribuida/python_concurrencia/ejercicio5.py
import threading

lock = threading.Condition()

class writer_thread(threading.Thread):
    def __init__(self, message):
        super().__init__()
        self.message = message

    
    def run(self):
        super().run()
        for i in range(5):
            lock.acquire()
            print(self.message, end="")
            lock.notify()
            if i < 4: lock.wait()
            lock.release()
        
t1 = writer_thread("Hola")
t2 = writer_thread(" Mundo\n")
t1.start()
t2.start()
t1.join()
t2.join()

            

